cells/prelude/library.nix

```diff
@@ -0,0 +1,92 @@
+{
+  inputs,
+  cell,
+}: let
+  inherit (inputs.nixpkgs) lib;
+  inherit (inputs.haskellNix.legacyPackages) haskell-nix;
+  inherit (haskell-nix) haskellLib;
+
+  default-compiler = "ghc8107";
+  extra-compilers = ["ghc924"];
+
+  # We clean-up src to avoid rebuild for unrelated changes:
+  src = haskellLib.cleanSourceWith {
+    src = inputs.self;
+    name = "cardano-prelude-src";
+    filter = path: type: let
+      relPath = lib.removePrefix "${inputs.self}/" path;
+    in
+      # excludes directories not part of cabal project:
+      (type
+        != "directory"
+        || (builtins.match ".*/.*" relPath != null)
+        || (!(lib.elem relPath [
+            "cells"
+          ])
+          && !(lib.hasPrefix "." relPath)))
+      # only keep cabal.project from files at root:
+      && (type == "directory" || builtins.match ".*/.*" relPath != null || (relPath == "cabal.project"))
+      && (lib.cleanSourceFilter path type)
+      && (haskell-nix.haskellSourceFilter path type)
+      && !(lib.hasSuffix ".gitignore" relPath)
+      # removes socket files
+      && lib.elem type ["regular" "directory" "symlink"];
+  };
+
+  projectConfig = {
+    pkgs,
+    config,
+    ...
+  }: {
+    name = "cardano-prelude";
+    inherit src;
+    # Uncomment for your system you want to see `nix flake show` without build agent for other systems:
+    #evalSystem = "x86_64-linux";
+    #evalSystem = "x86_64-darwin";
+    #evalSystem = "aarch64-darwin";
+    compiler-nix-name = lib.mkDefault default-compiler;
+    # Cross compilation support:
+    crossPlatforms = p:
+      lib.optionals pkgs.stdenv.hostPlatform.isLinux (
+        [
+          p.musl64
+        ]
+        ++ lib.optional pkgs.stdenv.hostPlatform.isx86_64
+        p.mingwW64
+      );
+    shell = {
+      withHoogle = false;
+      # Skip cross compilers for the shell
+      crossPlatforms = p: [];
+    };
+    modules = let
+      # deduce package names from the cabal project to avoid hard-coding them:
+      projectPackageNames =
+        builtins.attrNames (haskellLib.selectProjectPackages
+          (haskell-nix.cabalProject' (builtins.removeAttrs config ["modules"])).hsPkgs);
+    in [
+      {
+        packages =
+          lib.genAttrs projectPackageNames
+          (name: {configureFlags = ["--ghc-option=-Werror"];});
+      }
+    ];
+  };
+
+  # provide easy access to components, alternative compilers and a devshell module.
+  projectOverlays = with haskellLib.projectOverlays; [
```

Comment:
I _really_ find this stuff way harder to think about than a simple function `CompilerName -> (Project, Shell)` or whatever, but :shrug: